csharp/src/Ice/IceSSL/TransceiverI.cs

```diff
@@ -393,34 +406,97 @@ private bool startAuthenticate(IceInternal.AsyncCallback callback, object state)
             {
                 if(!_incoming)
                 {
-                    //
                     // Client authentication.
-                    //
-                    _writeResult = _sslStream.AuthenticateAsClientAsync(_host,
-                                                                        _instance.certs(),
-                                                                        _instance.protocols(),
-                                                                        _instance.checkCRL() > 0);
-                    _writeResult.ContinueWith(task => callback(state), TaskScheduler.Default);
+                    if (_instance.initializationData().clientAuthenticationOptions
+                        is SslClientAuthenticationOptions clientAuthenticationOptions)
+                    {
+                        _writeResult = _sslStream.AuthenticateAsClientAsync(clientAuthenticationOptions);
+                        _writeResult.ContinueWith(
+                            task =>
+                            {
+                                try
+                                {
+                                    _verified = task.IsCompletedSuccessfully;
+                                    if (_verified)
+                                    {
+                                        _cipher = _sslStream.CipherAlgorithm.ToString();
+                                        if (_sslStream.RemoteCertificate is X509Certificate2 remoteCertificate)
+                                        {
+                                            _certs = [remoteCertificate];
+                                        }
+                                    }
+                                    else
+                                    {
+                                        _certs = [];
+                                    }
+                                }
+                                finally
+                                {
+                                    callback(state);
+                                }
+                            },
+                            TaskScheduler.Default);
+                    }
+                    else
+                    {
+
+                        _writeResult = _sslStream.AuthenticateAsClientAsync(
+                            _host,
+                            _instance.certs(),
+                            _instance.protocols(),
+                            _instance.checkCRL() > 0);
```

Comment:
Can we build a SslClientAuthenticationOptions from this and use the same codepath for both cases.